### PR TITLE
Group Membership UI

### PR DESCRIPTION
**Pull request by dbkr**
* Group Member panel with support for inviting new group members and removing them from groups
 * Display group invites in the Invites section and accept / reject buttons in the GroupView
 * Make ConfirmUserAction dialog also support group members
 * Add `danger` prop to QuestionDialog (makes the primary button red)
 * Make MultiInviter support inviting to groups too

Requires https://github.com/matrix-org/matrix-js-sdk/pull/528
Requires https://github.com/vector-im/riot-web/pull/4830

### CODE REVIEW

**Comment by dbkr** [comment]
Test fails are due to needing the js-sdk branch (and also a test flake)

File: src/components/views/groups/GroupMemberList.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                fetching: false,
+            });
+        }).catch((e) => {
+            this.setState({fetching: false});
+            console.error("Failed to get group member list: " + e);
+        });
+    },
+
+    _createOverflowTile: function(overflowCount, totalCount) {
+        // For now we'll pretend this is any entity. It should probably be a separate tile.
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberList.js]: I'd rather just get this done as intended tbh

File: src/components/views/groups/GroupMemberList.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    },
+
+    makeGroupMemberTiles: function(query) {
+        const GroupMemberTile = sdk.getComponent("groups.GroupMemberTile");
+        query = (query || "").toLowerCase();
+
+        const memberList = this.state.members.filter((m) => {
+            if (query) {
+                //const matchesName = m.name.toLowerCase().indexOf(query) !== -1;
+                const matchesId = m.userId.toLowerCase().indexOf(query) !== -1;
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberList.js]: `.includes` ?

File: src/components/views/groups/GroupMemberList.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    onSearchQueryChanged: function(ev) {
+        this.setState({ searchQuery: ev.target.value });
+    },
+
+    makeGroupMemberTiles: function(query) {
+        const GroupMemberTile = sdk.getComponent("groups.GroupMemberTile");
+        query = (query || "").toLowerCase();
+
+        const memberList = this.state.members.filter((m) => {
+            if (query) {
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberList.js]: given `query` is constant for the `filter`, it'd be nicer to only apply the `filter` if `query`.

File: src/components/views/groups/GroupMemberList.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+            return true;
+        }).map((m) => {
+            return (
+                <GroupMemberTile key={m.userId} groupId={this.props.groupId} member={m} />
+            );
+        });
+
+        memberList.sort((a, b) => {
+            // should put admins at the top: we don't yet have that info
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberList.js]: nit: should be marked as `TODO`

File: src/components/views/groups/GroupMemberTile.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2017 Vector Creations Ltd
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberTile.js]: Copyright

File: src/utils/MultiInviter.js [hunk 6]
```diff
@@ -104,7 +115,14 @@ export default class MultiInviter {
             return;
         }
 
-        inviteToRoom(this.roomId, addr).then(() => {
+        let doInvite;
+        if (this.groupId !== null) {
+            doInvite = MatrixClientPeg.get().inviteUserToGroup(this.groupId, addr);
```
Comment (lukebarnard1) [inline src/utils/MultiInviter.js]: I was curious to know why we need `inviteToRoom` and it turns out that it's for inviting by email. I wonder if we're going to need to invite people to groups by email..

File: src/GroupInvite.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+export function showGroupInviteDialog(groupId) {
+    const UserPickerDialog = sdk.getComponent("dialogs.UserPickerDialog");
+    Modal.createTrackedDialog('Group Invite', '', UserPickerDialog, {
+        title: _t('Invite new group members'),
+        description: _t("Who would you like to add to this group?"),
+        placeholder: _t("Name or matrix ID"),
+        button: _t("Invite to Group"),
+        validAddressTypes: ['mx'],
+        onFinished: (shouldInvite, addrs) => {
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/GroupInvite.js]: I copied this style from elsewhere but ok

File: src/components/structures/GroupView.js [hunk 8]
```diff
@@ -295,6 +313,59 @@ export default React.createClass({
         }).done();
     },
 
+    _onAcceptInviteClick: function() {
+        this.setState({membershipBusy: true});
+        MatrixClientPeg.get().acceptGroupInvite(this.props.groupId).then(() => {
+            // don't reset membershipBusy here: wait for the membership change to come down the sync
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/GroupView.js]: We could certainly make higher level APIs that do this, yep, but right now this is just a wrapper around the REST API so I'd really like to keep it like that.

File: src/components/structures/GroupView.js [hunk 9]
```diff
@@ -371,6 +442,50 @@ export default React.createClass({
...
+                </div>;
+            }
+
+            return <div className="mx_GroupView_invitedSection">
+                {_t("%(inviter)s has invited you to join this group", {inviter: group.inviter.userId})}
+                <div className="mx_GroupView_membership_buttonContainer">
+                    <AccessibleButton className="mx_GroupView_textButton mx_RoomHeader_textButton"
+                        onClick={this._onAcceptInviteClick}
+                    >
+                        {_t('Accept')}
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/GroupView.js]: We should decide what quote style we're using for _t. I've been following our normal quote rules of preferring single quotes, but possibly we want a different one for _ts where you're more likely to use apostrophes.

File: src/components/structures/LoggedInView.js [hunk 10]
```diff
@@ -306,8 +306,9 @@ export default React.createClass({
             case PageTypes.GroupView:
                 page_element = <GroupView
                     groupId={this.props.currentGroupId}
+                    collapsedRhs={this.props.collapse_rhs}
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/structures/LoggedInView.js]: done

File: src/components/views/dialogs/ConfirmUserActionDialog.js [hunk 11]
```diff
@@ -91,17 +96,31 @@ export default React.createClass({
             );
         }
 
+        let avatar;
+        let name;
+        let userId;
+        if (this.props.member) {
+            avatar = <MemberAvatar member={this.props.member} width={48} height={48} />;
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/dialogs/ConfirmUserActionDialog.js]: It basically sounds like you're suggesting just getting the avatar url, userId and display name for either case and then using the BaseAvatar for both cases?

File: src/components/views/dialogs/ConfirmUserActionDialog.js [hunk 12]
```diff
@@ -91,17 +96,31 @@ export default React.createClass({
...
+            name = this.props.member.name;
+            userId = this.props.member.userId;
+        } else {
+            // we don't get this info from the API yet
+            avatar = <BaseAvatar name={this.props.groupMember.userId} width={36} height={36} />;
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/dialogs/ConfirmUserActionDialog.js]: oops

File: src/components/views/groups/GroupMemberInfo.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2017 Vector Creations Ltd
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+import GeminiScrollbar from 'react-gemini-scrollbar';
+
+
+module.exports = withMatrixClient(React.createClass({
+    displayName: 'GroupMemberInfo',
+
+    propTypes: {
+        matrixClient: PropTypes.object.isRequired,
+        groupId: PropTypes.string,
+        member: GroupMemberType,
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: I think I'd assumed it was implicit from the fact you're in a component called GroupMemberInfo. Done.

File: src/components/views/groups/GroupMemberInfo.js [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+        this.props.matrixClient.getGroupUsers(this.props.groupId).then((result) => {
+            this.setState({
+                members: result.chunk.map((apiMember) => {
+                    return groupMemberFromApiObject(apiMember);
+                }),
+                fetching: false,
+            });
+        }).catch((e) => {
+            this.setState({fetching: false});
+            console.error("Failed to get group member list: " + e);
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+        Modal.createDialog(ConfirmUserActionDialog, {
+            groupMember: this.props.member,
+            action: _t('Remove from group'),
+            danger: true,
+            onFinished: (proceed) => {
+                if (!proceed) return;
+
+                this.setState({removingUser: true});
+                this.props.matrixClient.removeUserFromGroup(this.props.groupId, this.props.member.userId).then(() => {
+                    dis.dispatch({
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+                    });
+                }).finally(() => {
+                    this.setState({removingUser: false});
+                });
+            },
+        });
+    },
+
+    _onCancel: function(e) {
+        dis.dispatch({
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    onRoomTileClick(roomId) {
+        dis.dispatch({
+            action: 'view_room',
+            room_id: roomId,
+        });
+    },
+
+    render: function() {
+        if (this.state.fetching || this.state.removingUser) {
+            const Loader = sdk.getComponent("elements.Spinner");
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: oh yes.

File: src/components/structures/GroupView.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (lukebarnard1) [reply src/components/structures/GroupView.js]: I continue to prefer `'` for all strings simply for the lack of pressing shift all the time. I say we use single quotes and just escape `\'` when we need them.

File: src/components/views/groups/GroupMemberInfo.js [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+            return <Loader />;
+        }
+        if (!this.state.members) return null;
+
+        let targetIsInGroup = false;
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+            giveModButton = <AccessibleButton className="mx_MemberInfo_field" onClick={this.onModToggle}>
+                {giveOpLabel}
+            </AccessibleButton>;*/
+        }
+
+        let adminTools;
+        if (kickButton || adminButton) {
+            adminTools =
+                <div>
+                    <h3>{_t("Admin tools")}</h3>
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberInfo.js [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+            <BaseAvatar name={this.props.member.userId} width={36} height={36} />
+        );
+
+        const memberName = this.props.member.userId;
+
+        const EmojiText = sdk.getComponent('elements.EmojiText');
+        return (
+            <div className="mx_MemberInfo">
+                <GeminiScrollbar autoshow={true}>
+                    <AccessibleButton className="mx_MemberInfo_cancel" onClick={this._onCancel}> <img src="img/cancel.svg" width="18" height="18"/></AccessibleButton>
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberList.js [hunk 22]
```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2017 Vector Creations Ltd.
```
[the remark this answers is not in the document]
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: done

File: src/components/views/groups/GroupMemberList.js (hunk 1 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: This is copied from MemberList and I don't really want to disappear down the rabbit hold of fixing that

File: src/components/views/groups/GroupMemberList.js (hunk 3 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: done

File: src/components/views/groups/GroupMemberList.js (hunk 2 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: done

File: src/components/views/groups/GroupMemberList.js (hunk 4 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: done

File: src/components/views/groups/GroupMemberTile.js (hunk 5 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberTile.js]: done

File: src/utils/MultiInviter.js (hunk 6 above)
Comment (dbkr) [reply src/utils/MultiInviter.js]: Maybe eventually :)

File: src/components/views/groups/GroupMemberInfo.js [hunk 23]
```diff
@@ -52,31 +53,32 @@ module.exports = withMatrixClient(React.createClass({
...
                 if (!proceed) return;
 
                 this.setState({removingUser: true});
-                this.props.matrixClient.removeUserFromGroup(this.props.groupId, this.props.member.userId).then(() => {
+                this.props.matrixClient.removeUserFromGroup(this.props.groupId, this.props.groupMember.userId).then(() => {
+                    // return to the user list
                     dis.dispatch({
                         action: "view_user",
-                        member: null,
+                        groupMember: null,
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberInfo.js]: I think this is a mistake

File: src/components/views/groups/GroupMemberInfo.js [hunk 24]
```diff
@@ -92,9 +94,10 @@ module.exports = withMatrixClient(React.createClass({
     },
 
     _onCancel: function(e) {
+        // Go back to the user list
         dis.dispatch({
             action: "view_user",
-            member: null,
+            groupMember: null,
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberInfo.js]: And this is a mistake

File: src/components/views/groups/GroupMemberInfo.js [hunk 25]
```diff
@@ -107,17 +110,14 @@ module.exports = withMatrixClient(React.createClass({
...
+        if (!this.state.groupMembers) return null;
 
-        let targetIsInGroup = false;
-        for (const m of this.state.members) {
-            if (m.userId == this.props.member.userId) {
-                targetIsInGroup = true;
-            }
-        }
+        const targetIsInGroup = this.state.groupMembers.some((m) => {
+            return m.userId == this.props.groupMember.userId;
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberInfo.js]: Redundant `==`, should be `===`

File: src/components/views/groups/GroupMemberList.js [hunk 26]
```diff
@@ -86,25 +87,28 @@ export default withMatrixClient(React.createClass({
         const GroupMemberTile = sdk.getComponent("groups.GroupMemberTile");
         query = (query || "").toLowerCase();
 
-        const memberList = this.state.members.filter((m) => {
-            if (query) {
+        let memberList = this.state.members;
+        if (query) {
+            memberList = memberList.filter((m) => {
                 //const matchesName = m.name.toLowerCase().indexOf(query) !== -1;
```
Comment (lukebarnard1) [inline src/components/views/groups/GroupMemberList.js]: Someone _might_ wonder why this is commented and it's not _super_ obvious that it's going to be implemented in future? Maybe a `TODO` ?

File: src/components/views/groups/GroupMemberInfo.js (hunk 23 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: oops, yes

File: src/components/views/groups/GroupMemberInfo.js (hunk 24 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: yep

File: src/components/views/groups/GroupMemberInfo.js (hunk 25 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberInfo.js]: done

File: src/components/views/groups/GroupMemberList.js (hunk 26 above)
Comment (dbkr) [reply src/components/views/groups/GroupMemberList.js]: fair :)

**Comment by lukebarnard1** [comment]
@richvdh, fresh test flake for you https://travis-ci.org/matrix-org/matrix-react-sdk/builds/277283447#L1101

**Pull request merged**